### ISSUE TITLE
Converted models are very slow

### ISSUE THREAD

**Issue by batrlatom** [tensorflow/tfjs#3808]
#### Describe the problem or feature request

I am trying to convert object detection models from TFHub, but I am unable to get good results. How can I get some meaningful  speedup?

For example, when I try to convert ssd_mobilenet_v2 model, I am getting something like 8s / image

`
tensorflowjs_converter --control_flow_v2=True --input_format=tf_hub --quantize_uint8=*  --strip_debug_ops=True --weight_shard_size_bytes=4194304 --output_node_names='Postprocessor/ExpandDims_1,Postprocessor/Slice' --signature_name 'serving_default' https://tfhub.dev/tensorflow/ssd_mobilenet_v2/2 test
`
Loaded with:

    ` const model_net = await tf.loadGraphModel("myserver.com/model.json", { fromTFHub: false })
`
The same model, but loaded directly from tfjs-models hub gets me speed around 1s/ image

     `const model_net = await tf.loadGraphModel("https://tfhub.dev/tensorflow/tfjs-model/ssd_mobilenet_v2/1/default/1", { fromTFHub: true })
`

#### TensorFlow.js version
1.7.4

#### TensorFlow version
TF cpu, 2.2.0

#### Browser version
react-native, android 10, xiaomi mi 6

**Comment by tafsiri** [comment]
@batrlatom Could you use [tf.profile](https://js.tensorflow.org/api/latest/#profile) with both your converted model **and** with the one from TFHub and upload the results here (as json files ideally). This will let us see which kernels are taking the most time and how that differs between the two.

Could you also upload the model.json of your converted model.

Thanks

**Comment by batrlatom** [comment]
Yes,
the converted models is at address: 
https://recall-models.s3.eu-central-1.amazonaws.com/ssd_mobilenet_v2/model.json

It was converted according this source: https://www.npmjs.com/package/@tensorflow-models/coco-ssd

`
tensorflowjs_converter --control_flow_v2=True --input_format=tf_hub --quantize_uint8=* --strip_debug_ops=True --weight_shard_size_bytes=4194304 --output_node_names='Postprocessor/ExpandDims_1,Postprocessor/Slice' --signature_name 'serving_default' https://tfhub.dev/tensorflow/ssd_mobilenet_v2/2 test
`
original model is at: 
https://tfhub.dev/tensorflow/ssd_mobilenet_v2/2 
My converted model has speed about 8s/im

original tfjs model from the same source:
https://tfhub.dev/tensorflow/tfjs-model/ssd_mobilenet_v2/1/default/1
This model has speed about 1s/im

I am going to try tf.profile,
Thanks

**Comment by pyu10055** [comment]
@batrlatom onnx model can be converted to TensorFlow saved model, you need to do following conversions:
pytourch => onnx => TF saved model => tfjs model

**Comment by batrlatom** [comment]
Hi, I have tried to convert codes from onnx and tf object detection model zoo for tf1 and tf2. 
I have a questions though.
In the model zoo for tf1, there are models both in the format of saved model and frozen inference graph. When I convert saved model, it has many time more nodes and edges (6800 ++ ) than model exported from frozen_inference graph ( 400 ) . Is the frozen model somehow optimized and layers fused? The models from tf2 model zoo do not have frozen model prepared for inference and after conversion has always high number of nodes. 

I also tried convert models from onnx to saved model and then to tfjs, but I am facing with even more nodes ( mainly batch norm nodes ) which I am unable to easily clean from the graph ( 14800 nodes for mobilenet v2? ) .  

How can I prune models from tf2 object detection model zoo or generally saved models?

**Comment by pyu10055** [comment]
@batrlatom do you know for those conversion from saved model, is training graph part of the conversion? I know the frozen inference graph specifically contains only the inference graph.

Our converter contains batch norm fusing logic, if you can share the saved models, we can see if the layout does not match the fusing logic.

**Comment by batrlatom** [comment]
Hi,
I have tried to convert this ssd_mobilenetv2 model from tf1 object detection zoo:
http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v2_coco_2018_03_29.tar.gz

There is saved model and frozen_inference_graph included.
When I try to export saved model, with this command 

`tensorflowjs_converter --input_format=tf_saved_model                        --output_format=tfjs_graph_model                        --output_node_names='Postprocessor/ExpandDims_1,Postprocessor/Slice'                        ./saved_model                        ./web_saved_model
`

the output is this ( 6340 nodes ):
2020-08-31 16:12:56.105180: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:845] Optimization results for grappler item: graph_to_optimize
  remapper: Graph size after: 6340 nodes (-12), 9937 edges (-12), time = 29.046ms.
  constant_folding: Graph size after: 6340 nodes (0), 9937 edges (0), time = 163.238ms.
  arithmetic_optimizer: Graph size after: 6340 nodes (0), 9937 edges (0), time = 141.29ms.
  dependency_optimizer: Graph size after: 6340 nodes (0), 9937 edges (0), time = 97.672ms.
  remapper: Graph size after: 6340 nodes (0), 9937 edges (0), time = 40.593ms.
  constant_folding: Graph size after: 6340 nodes (0), 9937 edges (0), time = 173.849ms.
  arithmetic_optimizer: Graph size after: 6340 nodes (0), 9937 edges (0), time = 155.553ms.
  dependency_optimizer: Graph size after: 6340 nodes (0), 9937 edges (0), time = 110.025ms.

Exported model can be found at : https://recall-models.s3.eu-central-1.amazonaws.com/tfjs_test/saved_model/model.json

******************************************************************************************************************************************
When I try to export frozen model with this code

`tensorflowjs_converter --input_format=tf_frozen_model                        --output_format=tfjs_graph_model                        --output_node_names='Postprocessor/ExpandDims_1,Postprocessor/Slice'                        ./frozen_inference_graph.pb                          ./web_frozen_model
`
, the output is this ( 602 nodes )
2020-08-31 16:11:25.779656: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:845] Optimization results for grappler item: graph_to_optimize
  remapper: Graph size after: 602 nodes (-12), 672 edges (-12), time = 16.65ms.
  constant_folding: Graph size after: 602 nodes (0), 672 edges (0), time = 31.037ms.
  arithmetic_optimizer: Graph size after: 602 nodes (0), 672 edges (0), time = 23.873ms.
  dependency_optimizer: Graph size after: 602 nodes (0), 672 edges (0), time = 6.672ms.
  remapper: Graph size after: 602 nodes (0), 672 edges (0), time = 4.222ms.
  constant_folding: Graph size after: 602 nodes (0), 672 edges (0), time = 24.182ms.
  arithmetic_optimizer: Graph size after: 602 nodes (0), 672 edges (0), time = 22.923ms.
  dependency_optimizer: Graph size after: 602 nodes (0), 672 edges (0), time = 7.057ms.

Exported model can be found at : https://recall-models.s3.eu-central-1.amazonaws.com/tfjs_test/frozen_model/model.json

It happens across all model zoo models where I am able to compare the output of export with saved model and frozen inference graph (tf1 only hae this)

**Comment by pyu10055** [comment]
@batrlatom The size difference could be related to that when converting the saved model the output_node_names param is not supported, which mean the sub graph after those nodes are included in the converted model. While in the frozen model, the subgraph after those specified output nodes is eliminated.

But if you look at the converted models, the weight bin files are pretty much the same, only the topology is large. But when you execute the model with specified node names, you can avoid executing the subgraph beyond the output nodes.

Can you help to benchmark the performance of those two output tfjs models? I suspect they are similar if you specify the same output nodes during the executeAsync() call.

please let us know what you have found out. thanks.

**Comment by batrlatom** [comment]
Hi, I was able to successfully convert ssd_resnet and ssd_mobilenet models and implement your postprocessing from tfjs-models. When I used ['StatefulPartitionedCall/Postprocessor/ExpandDims_1', 'StatefulPartitionedCall/Postprocessor/Slice'] as output nodes, it works quite well with webgl backend. But I have noticed one problem. When I want to use react-native, the performance is again extremely bad. On my xiaomi mi6 I can get 300ms on ssd_resnet from tf2 model zoo. The same code, same model yields 3000ms or more by using react-native. You can see my code bellow. Can you see where the problem could be? 

`<html>
<head>
    
    <script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs/dist/tf.min.js"></script>


    
    <script>
        // Notice there is no 'import' statement. 'tf' is available on the index-page


        function calculateMaxScores(
            scores, numBoxes,
            numClasses) {
            const maxes = [];
            const classes = [];
            for (let i = 0; i < numBoxes; i++) {
                let max = Number.MIN_VALUE;
                let index = -1;
                for (let j = 0; j < numClasses; j++) {
                    if (scores[i * numClasses + j] > max) {
                        max = scores[i * numClasses + j];
                        index = j;
                    }
                }
                maxes[i] = max;
                classes[i] = index;
            }
            return [maxes, classes];
        }

        tf.tidy(async () => {

            console.log('loading ...')
            //const model_net = await tf.loadGraphModel("https://recall-models.s3.eu-central-1.amazonaws.com/ssd_mobilenet_v2_fpnlite_320x320_q8/model.json", {fromTFHub: false})
            //const model_net = await tf.loadGraphModel('https://recall-models.s3.eu-central-1.amazonaws.com/ssd_mobilenet_v2_fpnlite_640x640_q8/model.json', {fromTFHub: false})
            //const model_net = await tf.loadGraphModel('https://recall-models.s3.eu-central-1.amazonaws.com/ssd_resnet_q8/model.json', {fromTFHub: false})
            const model_net = await tf.loadGraphModel('https://recall-models.s3.eu-central-1.amazonaws.com/ssd_racoon/model.json', {fromTFHub: false})
            console.log('backend: ' + tf.getBackend())

            console.log(model_net)

            var maxNumBoxes = 20
            var minScore = 0.5


            var i = 0;
            for (i = 0; i < 50; i++) {

                var start = new Date().getTime();
                console.log(i)
                //var vec = tf.randomNormal([1, 640, 640, 3], 0, 0, 'int32')
                var vec = tf.randomNormal([1, 320, 320, 3], 0, 0, 'int32')
                var result = await model_net.execute(vec, ['StatefulPartitionedCall/Postprocessor/ExpandDims_1', 'StatefulPartitionedCall/Postprocessor/Slice']);

                const height = vec.shape[1];
                const width = vec.shape[2];

                // 320x320
                //var boxes = result[0].reshape([1, 12804, 1, 4]);
                //var scores = result[1].reshape([1, 12804, 90]);

                //racoon
                var boxes = result[0].reshape([1, 1639, 1, 4]);
                var scores = result[1].reshape([1, 1639, 1]);
                //640x640
                //var boxes = result[0].reshape([1, 51150, 1, 4]);
                //var scores = result[1].reshape([1, 51150, 90]);



                const [maxScores, classes] =
                    calculateMaxScores(scores, scores.shape[1], scores.shape[2]);

                const prevBackend = tf.getBackend();

                tf.setBackend('cpu');

                const indexTensor = tf.tidy(() => {
                    var shapes = [boxes.shape[1], boxes.shape[3]]
                    boxes = boxes.dataSync()
                    const boxes2 = tf.tensor2d(boxes, shapes);

                    return tf.image.nonMaxSuppression(
                        boxes2, maxScores, maxNumBoxes, minScore, minScore);
                });

                 const indexes = indexTensor.dataSync();
                indexTensor.dispose();

                // restore previous backend
                tf.setBackend(prevBackend);

                var end = new Date().getTime();

                var time_dil = end - start;
                console.log(i + " :" + (time_dil));

                var node = document.createElement("LI");                 // Create a <li> node
                var textnode = document.createTextNode("" + time_dil);         // Create a text node
                node.appendChild(textnode);                              // Append the text to <li>
                document.getElementById("myList").appendChild(node);

            }

        });
    </script>
</head>

<body>
<div id="main">
    <ul id="myList"></ul>
</div>
</body>
</html>
`

**Comment by AzlinaZanariahMalik** [comment]
Hi I need help I'm new to tensorflow how do I freeze the model?

The one I use is save as saved_model.pb I'm using tensorflow 2.6 with Tensorflow object Detection API

My custom object Detection converted in tensorflowjs is always 12.7 mb and it lags and delay detecting on the web app but from the jupyter notebook it's not 

How can I optimize or compress into smaller size? Thanks
